php/libraries/LorisFormDictionaryImpl.class.inc

```diff
@@ -0,0 +1,184 @@
+<?php declare(strict_types=1);
+use \LORIS\Data\Dictionary\DictionaryItem;
+use \LORIS\Data\Dictionary\Category;
+use \LORIS\Data\Scope;
+use \LORIS\Data\Cardinality;
+/**
+ * Helper to extract a dictionary from a LorisForm, primarily
+ * used by instruments. This is based on lorisform_parser.php
+ *
+ * @license http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ */
+trait LorisFormDictionaryImpl
+{
+    /**
+     * Variable from NDB_Page. Must be included in trait to avoid
+     * PhanUndeclaredProperty errors.
+     */
+    public $page;
+
+    /**
+     * Variable from NDB_Page. Must be included in trait to avoid
+     * PhanUndeclaredProperty errors.
+     *
+     * @var \LorisForm
+     */
+    public $form;
+
+    /**
+     * Variable from NDB_BVL_Instrument. Must be included in trait
+     * to avoid PhanUndeclaredProperty errors.
+     */
+    public $testName;
+
+    /**
+     * Recursively extract non-group elements from $elements, and
+     * return the dictionary of the items.
+     *
+     * @param string   $prefix     The prefix for recursive groups
+     * @param Category $cat        The dictionary category to use for the
+     *                             items
+     * @param array    $elements   The LorisForm elements in the format
+     *                             of LorisForm->toElementArray().
+     * @param string   $groupLabel The quickform group being recursively
+     *                             extracted
+     *
+     * @return DictionaryItem[]
+     */
+    private function _getDataDictionaryGroup(
+        string $prefix,
+        \LORIS\Data\Dictionary\Category $cat,
+        array $elements,
+        string $groupLabel=''
+    ) : array {
+        $items = [];
+        $scope = new Scope(Scope::SESSION);
+        foreach ($elements AS $element) {
+            $card = new Cardinality(Cardinality::SINGLE);
+            if (!empty($element['label'])) {
+                $label = str_replace("&nbsp;", "", $element['label']);
+                $label = trim(preg_replace('/\s+/', ' ', $label));
+            } else {
+                $label = trim(preg_replace('/\s+/', ' ', $groupLabel));
+            }
+
+            if ($element['type'] === 'group') {
+                $items = array_merge(
+                    $items,
+                    $this->_getDataDictionaryGroup(
+                        $prefix,
+                        $cat,
+                        $element['elements'],
+                        $label,
+                    ),
+                );
+                continue;
+            }
```

Comment:
Unless I'm mistaken this handles tables in instruments? 

It seems like here you break it down and create the dictionary item for each cell in the table. that's okai and I think it's the status quo...

HOWEVER I always wondered if we could tackle this differently and this might be the occasion to do it.
The label for a cell in the table should be the concatenation of the column header + row label. do you think this is the right place to do it ?

Reply:
This is based on/done to be the same way as the lorisform_parser script in order to ensure that the old and new dictionaries are the same.

Maybe it would be a good idea to leave it this way initially, and then re-visit once #6936 is merged? That would make it easier to see the outcome of any changes.

Reply:
makes sense, note for the future